### DISCUSSION TITLE
Use the package react-gbajs

### DISCUSSION THREAD

**Discussion by macabeus** [city41/smaghetti#29] [General]
Hey, I just published the package `react-gbajs`: https://www.npmjs.com/package/react-gbajs 🎉 

I was looking to import it here, but I noticed that there are some differences between your [`/brute/gbajs`](/brute/gbajs) over the original version (for example https://github.com/city41/smaghetti/commit/86aa50fb35f7b13209bf0e09a8f054e7caf9f026).
And looks like that you are handling `canvas` in a different way, calling `setCanvasDirect` instead of just using a single `<canvas />`  element, which each one for a tile.

What do you think would need to be updated on the package to can use `react-gbajs` here?

**Comment by city41** [comment]
The brute directory is where I have node scripts for brute forcing things in the game for reverse engineering purposes. So it wouldn't use react-gbajs and can be ignored.

[here](https://github.com/city41/smaghetti/blob/main/src/components/LevelPlayer/GBAPlayer.tsx) is where I'd swap in react-gbajs. But in order to be able to use react-gbajs I would need to be able to maintain the scripting I added, the crash reporting and also not use requestAnimationFrame. Sadly, I think for emulators, setTimeout is the best way to time frames.

I still hack on gbajs and I have to admit I'm not sure I'm ready to give that up yet. Having the gba code in a node module will make it a black box and much harder to hack on. For sure when smaghetti is stable, react-gbajs would be great. For example this week I'm about to start building a tool for extracting large assets out of the game, and it will require digging into gbajs's memory data structures. I'm not sure yet what that entails, but I have a feeling I will need to add hooks into gbajs to get at that memory.

**Reply by city41** [reply]
I use `setCanvas` in the main gbajs in smaghetti. I think this is something that would be nice to change as well, if the canvas is large than 240x160, then gbajs scales the image manually in JS, which is very slow. `setCanvas`'s main purpose is to detect if that needs to happen and sets it up, where `setCanvasDirect` skips that check. 

It's much faster to use a 240x160 canvas and scale it with a css transform.

**Reply by macabeus** [reply]
Okay, I applied this suggestion on this PR: https://github.com/macabeus/react-gbajs/pull/5

**Reply by city41** [reply]
Nice. Don't get me wrong, I'm all for react-gbajs, I just don't think I'm quite ready to take it on yet. Like I said, I still need to hack on gbajs for more reverse engineering efforts. I know I will be building at least one more tool that needs gbajs internals. I do hope that reverse engineering is stabilizing somewhat soon and I won't need to muck with gbajs anymore 🤞

**Reply by macabeus** [reply]
There is no problem. Take your time.

I think curious that you are using gbajs to to all reverse engineering stuff and researches.
Like I said in my posts, I was doing these tasks using No$GBA and IDA. Is nice to see different approaches that other people apply to solve the same problems

**Reply by city41** [reply]
That's mainly because the tools I am building are either web based or node scripts. Scripting something like No$GBA would be tough, where as running gbajs in Node was really easy.

For example this is one of the tools

![image](https://user-images.githubusercontent.com/141159/120727606-f8a45080-c4a8-11eb-9d27-22ce4615de22.png)

**Reply by city41** [reply]
I've also been thinking I might need to have a custom gbajs in the long run. For example, not sure if you've played mario maker but it has a feature where it shows where you went in the level to help with item placement

![image](https://user-images.githubusercontent.com/141159/120728497-0f4ba700-c4ab-11eb-90d7-f30347691f2a.png)

I think that'd be cool to have, and I'd need to store mario's position and such as the user tests their level. Also was thinking of allowing the user to play the level with an invincible mario so they can just try stuff without worrying about dying.

**Reply by macabeus** [reply]
> it has a feature where it shows where you went in the level to help with item placement

It looks like a really nice feature, but I don't know how to add it being a game agnostic (and it's necessary to have it on `react-gbajs`).
Probably it would be possible if fetch the OAM state time by time and add them arbitrarily on the emulator.

> thinking of allowing the user to play the level with an invincible mario so they can just try stuff without worrying about dying.

I added this feature on `klo-gba.js`: https://github.com/macabeus/klo-gba.js/pull/99
To do it, I added support to freeze an address value.
If it's enough to have an invincible Mario, you can use it out of the box right now.

**Reply by macabeus** [reply]
Hey, one feature that you added on Smaghetti that I really would like to add on `klo-gba.js` is to show the real tiles, instead of just its ID and a pixel.
How did you get it?
